### PR TITLE
Byte counting at Congestion avoidance

### PR DESCRIPTION
**Pull request by junhochoi**
There is an issue in current congestion avoidance pseudocode
when `max_datagram_size * acked_packet.sent_bytes < congestion_window`
cwnd increment effectively becomes zero, especially because most
implementations use integer operation to calculate cwnd.

This happens when `acked_packet.sent_bytes` is a small value or
`congestion_window` is very big. It may lead to too conservative
cwnd growth during congestion avoidance.

To fix the issue, I propose to use a TCP ABC style cwnd growth
during congestion avoidance: https://www.rfc-editor.org/rfc/rfc3465.html#section-2.1

Basically it stores accumulated acked bytes in a new
variable `bytes_acked` and increase cwnd by `max_datagram_size` when
`bytes_acked` reaches to cwnd.

### CODE REVIEW

**Comment by ianswett** [comment]
I'm not sure this is necessary, since we don't specify the variables as integers, but if we were going to make a change, I'd prefer the CWND increment in bytes rather than max_datagram_size, since it's a smoother increase.

**Comment by junhochoi** [comment]
> I'm not sure this is necessary, since we don't specify the variables as integers, but if we were going to make a change, I'd prefer the CWND increment in bytes rather than max_datagram_size, since it's a smoother increase.

If we increase CWND in bytes (let say 10 x 1/10 mss, rather than 1 full mss), is it resulting too many small sending packets?

**Comment by janaiyengar** [comment]
I agree with @ianswett : I don't think this is necessary... perhaps a comment might help clarify that implementers should be careful about integer division here?

FWIW, you'll also need to multiply bytes_acked in OnCongestionEvent() by 0.5, and reset it to 0 on a persistent congestion event.

**Comment by kazu-yamamoto** [comment]
I would like to add comments in the following places:
- `/` in `OnPacketAcked`
- `congestion_window *= kLossReductionFactor` in `CongestionEvent`
- "`lost_packets` should be retransmitted" in `OnPacketsLost`
- "`max_ack_delay` is updated when TP exchanged, default to 25ms" in `Initialization`
- "A sender can send bytes of `congestion_window` - `bytes_in_flight`" in `Variables of interest`

They might be trivial for experts but it is difficult for beginners like me to understand the assumptions.

**Comment by kazu-yamamoto** [comment]
@janaiyengar This is an off topic but there is no `OnCongestionEvent`. The current name is `CongestionEvent`. I would like to suggest to change it to `OnCongestionEvent` or `OnNewCongestionEvent`.

**Comment by ianswett** [comment]
@kazu-yamamoto renaming to OnCongestionEvent seems fine.  

If you want to make a PR, we can go through the comments you want to add, but "lost_packets should be retransmitted" is potentially misleading, since packets aren't retransmitted and their frames aren't always retransmitted.  Also, the max_ack_delay is 0, not 25ms, until the transport parameters are received.

**Comment by junhochoi** [comment]
@janaiyengar commenting on integer division is also good I think, but this PR basically removes such concern by not using division any more.

I looked at other places using a division in the pseudo code but other than Pacing section it uses a constant, so improving here is not a bad thing I think.

Also I updated to reduce bytes_acked in CongestionEvent() and reset persistent congestion event.

**Comment by kazu-yamamoto** [comment]
> If you want to make a PR, we can go through the comments you want to add, but "lost_packets should be retransmitted" is potentially misleading, since packets aren't retransmitted and their frames aren't always retransmitted.

Yes. What about this?  "The information in `lost_packets` should be retransmitted if necessary."

> Also, the max_ack_delay is 0, not 25ms, until the transport parameters are received.

I meant that it is initialized to 0 and updated when  the transport parameters are received. If the parameter is not specified, 25ms is set to this parameter.

**Comment by ianswett** [comment]
@kazu-yamamoto those changes sound good to me, but it'd be good to see them in context.  Want to send me a PR?

**Comment by kazu-yamamoto** [comment]
Yes. I will do it on next Monday.

File: draft-ietf-quic-recovery.md [hunk 1]
```diff
@@ -1483,6 +1483,11 @@ ssthresh:
   the mode is slow start and the window grows by the number of bytes
   acknowledged.
 
+bytes_acked:
+: Store how many bytes acked so far. During congestion avoidance,
+  congestion window will be incremented by max_datagram_size
+  when this value is same or bigger than the congestion window.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md:1489]:
```suggestion
: Records number of bytes acked and used during congestion
  avoidance, for incrementing the congestion window.
```

File: draft-ietf-quic-recovery.md [hunk 2]
```diff
@@ -1562,6 +1569,7 @@ window.
        congestion_window *= kLossReductionFactor
        congestion_window = max(congestion_window, kMinimumWindow)
        ssthresh = congestion_window
+       bytes_acked *= kLossReductionFactor
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md:1572]: move this up to the line below where the congestion window is reduced

File: draft-ietf-quic-recovery.md [hunk 3]
```diff
@@ -1541,9 +1547,10 @@ newly acked_packets from sent_packets.
       congestion_window += acked_packet.sent_bytes
     else:
       // Congestion avoidance.
-      congestion_window +=
-        max_datagram_size * acked_packet.sent_bytes
-        / congestion_window
+      bytes_acked += acked_packet.sent_bytes
+      if (bytes_acked >= congestion_window):
+        bytes_acked -= congestion_window
+        congestion_window += max_datagram_size
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md:1553]: There's a minor behavior difference here from the old code, in that the cwnd here increases by at most 1. I think that's fine; I don't think it's possible to have an increase that is larger. The following code is the same as the old behavior, but I don't think it's necessary.

```suggestion
      while (bytes_acked >= congestion_window):
        bytes_acked -= congestion_window
        congestion_window += max_datagram_size
```

File: draft-ietf-quic-recovery.md (hunk 1 above)
Comment (junhochoi) [reply draft-ietf-quic-recovery.md]: LGTM. Updated.

File: draft-ietf-quic-recovery.md (hunk 2 above)
Comment (junhochoi) [reply draft-ietf-quic-recovery.md]: Done. Thanks.

File: draft-ietf-quic-recovery.md (hunk 3 above)
Comment (junhochoi) [reply draft-ietf-quic-recovery.md:1552]: I don't think `while()` is necessary because `OnPacketAcked()` (recently changed) is called per acked packet, so `acked_packet.sent_bytes` is always smaller than `max_datagram_size`.

Comment (kazu-yamamoto) [reply draft-ietf-quic-recovery.md:1552]: I agree with @junhochoi.

**Comment by kazu-yamamoto** [comment]
I will implement this today for review. Please wait for a while.

File: draft-ietf-quic-recovery.md [hunk 4]
```diff
@@ -1483,6 +1483,10 @@ ssthresh:
   the mode is slow start and the window grows by the number of bytes
   acknowledged.
 
+bytes_acked:
+: Records number of bytes acked and used during congestion
+  avoidance, for incrementing the congestion window.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md:1488]:
```suggestion
: Records number of bytes acked, and used for incrementing
  the congestion window during congestion avoidance.
```

File: draft-ietf-quic-recovery.md (hunk 3 above)
Comment (janaiyengar) [reply draft-ietf-quic-recovery.md:1552]: Agreed. I wanted to note the difference, but as I said, I don't think it's necessary.

**Comment by janaiyengar** [comment]
FWIW, I should add that this is how quicly implements it. If I remember correctly, that it's also how Chrome implements it.

File: draft-ietf-quic-recovery.md [hunk 5]
```diff
@@ -1483,6 +1483,10 @@ ssthresh:
   the mode is slow start and the window grows by the number of bytes
   acknowledged.
 
+bytes_acked:
+: Records number of bytes acked, and used for incrementing
```
Comment (ianswett) [inline draft-ietf-quic-recovery.md:1487]:
```suggestion
:  Number of bytes acknowledged which have not yet been used to increment
```

File: draft-ietf-quic-recovery.md [hunk 6]
```diff
@@ -1560,6 +1566,7 @@ window.
      if (!InCongestionRecovery(sent_time)):
        congestion_recovery_start_time = now()
        congestion_window *= kLossReductionFactor
+       bytes_acked *= kLossReductionFactor
```
Comment (ianswett) [inline draft-ietf-quic-recovery.md:1569]: I thought this would be set to 0 here.  Why reduce it?

Comment (janaiyengar) [reply draft-ietf-quic-recovery.md:1569]: The current cwnd is (cwnd + bytes_acked/cwnd). Multiplying that by kLossReductionFactor means you multiply bytes_acked by kLossReductionFactor as well.

Comment (junhochoi) [reply draft-ietf-quic-recovery.md:1569]: Just thinking again... this is when a recovery episode starts and we enter congestion avoidance by ssthresh=cwnd. And bytes_acked is only used in congestion avoidance and doing nothing during slow start.

I think we simply clear bytes_acked to 0 as @ianswett said because it's a new start of congestion avoidance, or no need to touch bytes_acked because even if there is a previous leftover bytes_acked, it will not impact much.

Comment (ianswett) [reply draft-ietf-quic-recovery.md:1569]: Thanks for the explanation.  I think I'm fine either way, but questions like this are one of the reason I liked the old approach of assuming it's the implementation's responsibility to deal with rounding/etc.

Reading this again, I realize there's a somewhat substantive difference in behavior between the previous pseudocode and this PR, because the previous pseudocode would increment CWND in bytes, not only max_datagram_size.

Comment (janaiyengar) [reply draft-ietf-quic-recovery.md:1569]: This isn't a big difference, so I'm fine with any resolution here. @ianswett : I don't understand what you mean by the substantive difference. The previous code incremented by (bytes_acked/cwnd) fraction of max_datagram_size on each ack, this one accumulates that fraction until it is 1 and then adds max_datagram_size. It is functionally the same.

Comment (ianswett) [reply draft-ietf-quic-recovery.md:1569]: I agree that this is a small difference either way.

File: draft-ietf-quic-recovery.md [hunk 7]
```diff
@@ -1541,9 +1546,10 @@ newly acked_packets from sent_packets.
       congestion_window += acked_packet.sent_bytes
     else:
       // Congestion avoidance.
-      congestion_window +=
-        max_datagram_size * acked_packet.sent_bytes
-        / congestion_window
+      bytes_acked += acked_packet.sent_bytes
+      if (bytes_acked >= congestion_window):
+        bytes_acked -= congestion_window
+        congestion_window += max_datagram_size
```
Comment (ianswett) [inline draft-ietf-quic-recovery.md:1552]: This part contains the larger change.  Previously, congestion_window was adjusted in 1 byte increments.  Now it's adjusted in max_datagram_size increments.  In practice, it likely doesn't matter that much, but I am concerned about perpetuating packet-size quantization in congestion control, since my impression is that it's an artifact of the Linux kernel, not an intentional design choice.

I believe BSD-based OS's use bytes instead of packets, and bytes always seemed more correct to me.

However, this isn't that difficult to fix.  How about:
  cwnd_in_packets = congestion_window / max_datagram_size
  congestion_window += bytes_acked / cwnd_in_packets
  bytes_acked = bytes_acked % cwnd_in_packets

Comment (janaiyengar) [reply draft-ietf-quic-recovery.md:1552]: I don't see how this is packet-based computation. Yes, the cwnd is incremented only when it can be incremented in max_datagram_size chunks, but at any given time, you know exactly what your fractional component is: (bytes_acked/congestion_window). At any given point in time, your total send allowance, in bytes, is (congestion_window + bytes_acked * max_datagram_size / congestion_window).

This fractional difference only matters if you are sending packets smaller than max_datagram_size.

Comment (junhochoi) [reply draft-ietf-quic-recovery.md:1552]: FYI, FreeBSD uses byte counting by default and use the same mechanism of this PR: https://github.com/freebsd/freebsd/blob/0b7e58eb670d67cdc0268be5d588839038aa5abf/sys/netinet/tcp_input.c#L355

And I think @ianswett 's code has a problem
> bytes_acked = bytes_acked % cwnd_in_packets
this will make bytes_acked too small (only between 0 - cwnd_in_packets - 1 range)

I'm happy with the current logic.

Comment (martinthomson) [reply draft-ietf-quic-recovery.md:1552]: > This fractional difference only matters if you are sending packets smaller than max_datagram_size.

This isn't quite right.  Based on this, you need to receive a complete window before you increase the window size.  Previously, you could maybe increment by smaller amounts sooner.  The difference is between a continuous increase (or at least an increase with every acknowledgments) and a stepped increase.  Let's say max_datagram_size is 1000 and your window is currently 10,000, you will have to receive acknowledgements for 10 packets before you can go to 11,000.  If you increase progressively, then you might go to 10,100 then 10,200, and so forth sooner.

If the assertion is that you can't use that small extra amount, consider that the window could be 9,999 to begin with.  That is entirely likely as you rarely settle on a strict multiple of packet size.  The first arriving acknowledgment would allow another full-sized packet to be sent much sooner than otherwise.

So this is a functional change, even if it you might easily decide that it is acceptable.

Comment (martinthomson) [reply draft-ietf-quic-recovery.md:1552]: Oh, I should probably concede that more increases to the congestion window will result in the final congestion window being smaller, so you are trading a larger final increase (this PR) with earlier availability of bytes (status quo).

Say you get 10 acks for 1000 bytes, with a starting cwnd of 10,000 and max_datagram_sizeo of 1000.

With this code, you end up with no change until the last ack, where you go to 11,000.

With the previous code, you would go to: 10100, 10199, 10297, 10394, 10490, 10585, 10680, 10773, 10866, 10958.  (That's using double-precision logic, you lose two bytes if you round.)

That's less in the end, but the value of that to you depends on whether you can use any of those intermediate values.

Comment (martinthomson) [reply draft-ietf-quic-recovery.md:1552]: If you want to realize increments immediately, but not lose any credit, then I think that the algorithm is:

```
credit = saved_credit + (bytes_acked * max_datagram_size)
congestion_window += credit / congestion_window
saved_credit = credit % congestion_window
```

This uses integer math to produce a value that matches the double-precision logic from my previous comment.

Comment (ianswett) [reply draft-ietf-quic-recovery.md:1552]: Yesterday was a long day, so I suspect my pseudocode has an error, but as @mt points out this does change how the pseudocode works.

Comment (junhochoi) [reply draft-ietf-quic-recovery.md:1552]: In the scenario @martinthomson mentioned, I think the problem with small increase is that it can result in a small sized packets.

e.g. bytes_in_flight = 10000, ssthresh = cwnd = 10000, mss = 1000
  -> received ack of 1000 bytes
  -> byts_in_flight = 9000, cwnd = 10100

Now sender can send 1100 bytes, can result in sending 2 packets, with 1000 bytes and 100 bytes each when there is a enough size of data is ready. The pseudocode in this PR doesn't have such issue.

Comment (junhochoi) [reply draft-ietf-quic-recovery.md:1552]: Briefly tested using quiche (https://github.com/cloudflare/quiche/pull/610) -- when downloading 10Mbytes file using 10Mbit/40ms link, total # of sender packets can be reduced by 14% (mostly small packets than MSS) with slightly better performance.

Comment (ianswett) [reply draft-ietf-quic-recovery.md:1552]: The small fragmented packet issue is an interesting one, is there an existing RFC which discusses this we could cite with advice on how to avoid overly small packets?

Comment (kazu-yamamoto) [reply draft-ietf-quic-recovery.md:1552]: What about RFC 3465?

Comment (junhochoi) [reply draft-ietf-quic-recovery.md:1552]: This mechanism described in RFC3465 as in the original PR description but the RFC itself has no pseudocode. I don't know other references discussing about those small fragments due to byte counting in congestion avoidance -- probably it's because most of TCP RFC use window-based cwnd, not byte based?

**Comment by junhochoi** [comment]
neqo is implmenting this logic too https://github.com/mozilla/neqo/pull/875

File: draft-ietf-quic-recovery.md (hunk 7 above)
Comment (janaiyengar) [reply draft-ietf-quic-recovery.md:1552]: Yes, this is a change in how the sender works, but it's not a significant difference. And it is in the pseudo code in the appendix, not in the text. 

Let's be clear that we are defining something that we are calling Reno for QUIC, knowing fully well that it's just one of many controllers that people will attempt. The current pseudo code is not a reference, it is one way of doing Reno. What this PR proposes is what a number of implementations already do.

I don't think there's much more to see here. I'm fine with not including the PR, or with including the PR. Most implementations are likely to implement it the way the PR does, but the current pseudo code is also good in that it is algorithmically clear.

Comment (ianswett) [reply draft-ietf-quic-recovery.md:1552]: I think this PR is fine, but I'm also not seeing a big reason for changing the current pseudocode.

It turns out Chrome doesn't do it this way anymore, but we used to.

**Comment by Blandine11** [comment]
I would like to ask a question about quiche's congestion control algorithms.
Is there an algorithm that uses ECN marking to detect or control congestion?

**Comment by junhochoi** [comment]
> I would like to ask a question about quiche's congestion control algorithms.
> Is there an algorithm that uses ECN marking to detect or control congestion?

@Blandine11 If you have a quiche-specific question please open a new ticket on https://github.com/cloudflare/quiche/issues.

File: draft-ietf-quic-recovery.md (hunk 7 above)
Comment (junhochoi) [reply draft-ietf-quic-recovery.md:1552]: I implemented ABC for cubic in quiche https://github.com/cloudflare/quiche/pull/624 and did some evaluation:
Measured using quiche apps (quiche-server and quiche-client) in 5mbps/60ms rtt/0% loss. Also last highlighted row is when I changed MAX_DATAGRAM_SIZE to 1350 to match with what quiche apps use (quiche has MAX_DATAGRAM_SIZE constant). 

Screen Shot 2020-08-28 at 5 21 49 PM

You can see it can save some amount of sent packets (3.8% for cubic, 6.7% for reno, 9.3% for cubic and 1350 mss aligned) with slightly faster download time.

I think it's up to editors for deciding to merge this PR not, but I believe it's good for implementers to start with less issue.

Comment (janaiyengar) [reply draft-ietf-quic-recovery.md]: I'm surprised to see these differences, @junhochoi -- why are the lost packets so different? Why's there any loss when the network has 0% loss?

Comment (junhochoi) [reply draft-ietf-quic-recovery.md]: @janaiyengar it's basically loss by congestion. e.g. https://blog-cloudflare-com-assets.storage.googleapis.com/2020/05/cubic-hs.png (cubic+hystart case). All values are median number over 10 run. And lost packet here is the number of lost packets detected by quiche sender, so actual packet loss may be lower than that.

Comment (janaiyengar) [reply draft-ietf-quic-recovery.md]: Are these controlled at the bottleneck or is this uncontrolled loss on the Internet? The number of packets lost seem quite different, and I'm trying to understand why those loss numbers are so different.

Comment (junhochoi) [reply draft-ietf-quic-recovery.md]: This is a simulated environment, not an public internet. In this testing I used MacOS dummynet. As I said this number is what recovery module detected as loss (which will trigger a congestion event). So may not be an actual loss (e.g. delivered late after a time threshold). I didn't measure actual packet loss.
What's your expectation?

**Comment by kazu-yamamoto** [comment]
So, how to decide including or declining this PR?

**Comment by janaiyengar** [comment]
The text currently says that the congestion window MUST NOT increase by more than 1 max datagram size per acknowledged cwnd. The pseudo code currently reflects what the text says. In that sense, the pseudo code is correct.

This PR is an alternative implementation of the text. As such, it doesn't fix a problem in the text, it fixes a problem with implementations that use integers.

The pushback on this PR is that the behavior is different than the existing pseudo code. Broadly, I agree with that, though, as I have said earlier, it shouldn't matter in practice. That said, the difference in behavior exists. The pseudo code in the draft is meant as clarification of the text, and implementations will and do differ from the pseudo code in various places.

My suggestion then would be to drop this PR, and instead add a comment above the concerning line saying something like: "Implementers that use an integer representation for the congestion window should be careful with the following division, and can use the alternative approach suggested in Section 2.1 of RFC 3465."

**Comment by junhochoi** [comment]
> My suggestion then would be to drop this PR, and instead add a comment above the concerning line saying something like: "Implementers that use an integer representation for the congestion window should be careful with the following division, and can use the alternative approach suggested in Section 2.1 of RFC 3465."

To clarify, close this one and open a new PR about text changes only? I'm ok to do that.

**Comment by junhochoi** [comment]
Opened #4082 for text changes.
